src/rest_api/api/search/superstructure_search.rs

```diff
@@ -0,0 +1,177 @@
+        }
+        None => None,
+    };
+
+    let results = superstructure_search(
+        &searcher,
+        &query_canon_taut,
+        &matching_scaffolds,
+        fingerprint.0.as_bitslice(),
+        &descriptors,
+        result_limit,
+        extra_query,
+    );
+
+    let mut results = match results {
+        Ok(results) => results,
+        Err(e) => {
+            return GetStructureSearchResponse::Err(Json(StructureResponseError {
+                error: e.to_string(),
+            }))
+        }
+    };
+
+    let mut used_tautomers = false;
+    let mut num_tauts_used = 0;
+    if !results.is_empty() {
+        num_tauts_used = 1;
+    }
+
+    if results.len() < result_limit {
+        let tautomers = get_tautomers(&query_canon_taut);
+
+        if tautomers.len() > 1 && tautomer_limit > 1 {
+            for test_taut in tautomers {
+                // don't reuse the canonical tautomer
+                if test_taut.as_smiles() == query_canon_taut.as_smiles() {
+                    continue;
+                }
+
+                let taut_attributes = get_cpd_properties(&test_taut);
+
+                let taut_attributes = match taut_attributes {
+                    Ok(taut_attributes) => taut_attributes,
+                    Err(_) => continue,
+                };
+
+                let (taut_fingerprint, taut_descriptors) = taut_attributes;
+
+                let matching_scaffolds = match scaffolds {
+                    Some(scaffolds) => {
+                        let scaffold_matches = scaffold_search(&test_taut, scaffolds);
+                        match scaffold_matches {
+                            Ok(scaffold_matches) => Some(scaffold_matches),
+                            Err(e) => {
+                                return GetStructureSearchResponse::Err(Json(
+                                    StructureResponseError {
+                                        error: e.to_string(),
+                                    },
+                                ))
+                            }
+                        }
+                    }
+                    None => None,
+                };
+
+                let taut_results = superstructure_search(
+                    &searcher,
+                    &test_taut,
+                    &matching_scaffolds,
+                    taut_fingerprint.0.as_bitslice(),
+                    &taut_descriptors,
+                    result_limit,
+                    extra_query,
+                );
+
+                let taut_results = match taut_results {
+                    Ok(taut_results) => taut_results,
+                    Err(_) => continue,
+                };
+
+                results.extend(&taut_results);
+                num_tauts_used += 1;
+
+                if !used_tautomers {
+                    used_tautomers = true;
+                }
+
+                if results.len() > result_limit || num_tauts_used == tautomer_limit {
+                    break;
```

Comment:
Break is similarly something that could be captured with a `filter` or `take` or something else. The more I write about this the more I'm OK with the monster for loop 😄

Reply:
But a monster for loop does lose out on the ability to do `par_iter` for CPU-based parallelism in the future.

Reply:
The `par_iter` idea is interesting. I guess it would depend on how often and how many tautomers are used. If the user asks for 1000 tautomers (maybe for superstructure search of a really complicated compound), then I could imagine `par_iter` being super helpful. I imagine this would be an atypical ask though. 

That said, we could also just default to `par_iter` and test the requested number of tautomers and then just return up to the `result_limit` to be comprehensive and safe in the case of a large `tautomer_limit`.

Reply:
Alright, I bit the bullet and went for the `par_iter`. Now we will use the full specified `tautomer_limit`. Also, we should default to `tautomer_limit = 0` to speed things up.